### DISCUSSION TITLE
Questions about contracts from Lesson 3

### DISCUSSION THREAD

**Discussion by 0xwtfwtf** [Cyfrin/foundry-full-course-cu#294] [Q&A] [answered]
Below I have pasted the 2 contracts we created for reference (the same ones used in Lessons 2 and 3).

1) How does the sfStore function know that _newSimpleStorageNumber is supposed to contain favoriteNumber from the SimpleStorage.sol contract? Also, why doesn't StorageFactory also return names since that was also created and stored by SimpleStorage?

2) In the sfGet function, why are we only returning _simpleStorageIndex and not also _newSimpleStorageNumber? 

3) And, in sfGet, does "SimpleStorage" refer to the contract and mySimpleStorage is a variable? If mySimpleStorage is a variable, why is it not created in the parentheses together with _simpleStorageIndex and _newSimpleStorageNumber? 

4) Does anyone have any suggested exercises for me to practice these concepts in Lessons 2 and 3?

Contract: SimpleStorage
```javascript
contract SimpleStorage {
    
    uint256 internal myFavoriteNumber;
       
    struct Person{
        uint256 favoriteNumber;
        string name;
    } 

    Person[] public listOfPeople; 


    function addPerson(string memory _name, uint256 _favoriteNumber) public {
        listOfPeople.push( Person(_favoriteNumber, _name) );
        nameToFavoriteNumber[_name] = _favoriteNumber; 
    }

    mapping(string => uint256) public nameToFavoriteNumber; 


    function store(uint256 _favoriteNumber) public {
        myFavoriteNumber = _favoriteNumber;
    } 
   
    function retrieve() public view returns(uint256){
        return myFavoriteNumber; 
    }


   }
```

Contract StorageFactory

```javascript
import {SimpleStorage} from "./simplestorage.sol";

contract StorageFactory{

    SimpleStorage[] public listOfSimpleStorageContracts; 

    function createSimpleStorageContract() public {
        SimpleStorage newSimpleStorageContract = new SimpleStorage();
        listOfSimpleStorageContracts.push(newSimpleStorageContract);
    }

    function sfStore(uint256 _simpleStorageIndex, uint256 _newSimpleStorageNumber) public{
        SimpleStorage mySimpleStorage = listOfSimpleStorageContracts[_simpleStorageIndex];
        mySimpleStorage.store(_newSimpleStorageNumber);
    }


    function sfGet(uint256 _simpleStorageIndex) public view returns(uint256){
        SimpleStorage mySimpleStorage = listOfSimpleStorageContracts[_simpleStorageIndex].retrieve;
        
    }


}
```

**Comment by PatrickAlphaC** [comment, marked as answer]
> How does the sfStore function know that _newSimpleStorageNumber is supposed to contain favoriteNumber from the SimpleStorage.sol contract? 

The `_newSimpleStorageNumber` doesn't `contain` the `favoriteNumber` in SimpleStorage.sol, it calls the `store` function on our SimpleStorage.sol, so in fact, it adds it! 

> Also, why doesn't StorageFactory also return names since that was also created and stored by SimpleStorage?

We can make our StorageFactory do whatever we want! And we decided we wanted our factory to _only_:
1. Create simple storage contracts via `createSimpleStorageContract`
2. Store numbers in those contracts via `sfStore`
3. Read numbers from the via `sfGet`

If you want, you can challenge yourself and try to get the factory to also return names!

> In the sfGet function, why are we only returning _simpleStorageIndex and not also _newSimpleStorageNumber?

In `sfGet`, we are returning the`myFavoriteNumber` in the respective SimpleStorage. In `sfGet`, we are _retreiving_ numbers that have already been stored in our SimpleStorage contracts. We are not adding any new ones. When we call `sfStore` with a `_newSimpleStorageNumber`, that gets stored in a `myFavoriteNumber`. Think of it like this:

Factory starts with 0 SimpleStorage Contracts, 
```
listOfSimpleStorageContracts: []
```

then we call `createSimpleStorageContract`, and now we have:
```
listOfSimpleStorageContracts: [newSimpleStorageContract1]
```

If we called `createSimpleStorageContract` again, we'd create 2 _different_ SimpleStorage contracts

```
listOfSimpleStorageContracts: [newSimpleStorageContract0, newSimpleStorageContract1]
```

Each SimpleStorage contract has different favorite numbers:

```
newSimpleStorageContract0.myFavoriteNumber = 0

newSimpleStorageContract1.myFavoriteNumber = 0
```

When we call `sfStore` on our factory, we pick _both_ the simplestorage contract & the new favorite number.

For example, if we then call `sfStore(0, 15)` we'd then have:

```
newSimpleStorageContract0.myFavoriteNumber = 15

newSimpleStorageContract1.myFavoriteNumber = 0
```

Then we call `sfStore(1, 33)` we'd then have:
```
newSimpleStorageContract0.myFavoriteNumber = 15

newSimpleStorageContract1.myFavoriteNumber = 33
```

Then we call `sfStore(1, 7)` we'd then have:
```
newSimpleStorageContract0.myFavoriteNumber = 15

newSimpleStorageContract1.myFavoriteNumber = 7
```

> And, in sfGet, does "SimpleStorage" refer to the contract and mySimpleStorage is a variable? If mySimpleStorage is a variable, why is it not created in the parentheses together with _simpleStorageIndex and _newSimpleStorageNumber?

I think I answered this above? 

> Does anyone have any suggested exercises for me to practice these concepts in Lessons 2 and 3?

Yes! In remix, start calling a bunch of the functions, BUT before you call each function, try to guess what the output will be, and how it will affect the other functions.
